MREUnityRuntime/MREUnityRuntimeLib/Animation/Interpolations.cs

```diff
@@ -0,0 +1,180 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+using CubicBezier = MixedRealityExtension.Util.CubicBezier;
+using Newtonsoft.Json.Linq;
+using Quaternion = UnityEngine.Quaternion;
+using UnityMath = UnityEngine.Mathf;
+
+namespace MixedRealityExtension.Animation
+{
+	internal static class Interpolations
+	{
+		private static Quaternion tempA;
+		private static Quaternion tempB;
+		private static Quaternion tempMix;
+
+		internal static void Interpolate(JToken a, JToken b, float linearT, ref JToken mix, CubicBezier easing)
+		{
+			var easedT = easing.Sample(linearT);
+
+			// compound types
+			if (a.Type == JTokenType.Object)
+			{
+				JObject A = (JObject)a;
+				JObject B = (JObject)b;
+				JObject Mix = (JObject)mix;
+
+				if (A.ContainsKey("x") && A.ContainsKey("y"))
```

Comment:
The interpolator should not have to inspect the value and rediscover how to interpolate its value every frame. I recommend making a set of interpolator functions, one for each type: QuaternionInterpolator, Vector3Interpolator, ValueInterpolator, etc. It should be possible using targetpath to know ahead of time which interpolator to use, then cache that interpolator method somewhere appropriate, perhaps in the keyframe or on the track.